[ext-docs] Reduces friction for CUJ "I want to know how to use the Storage API" (#3961)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/storage/index.md

{% endDetails %}

### Storage areas

The Storage API is divided into the following four buckets ("storage areas"): 

- [**storage.local**][prop-local]
  - Data is stored locally, which is cleared when the extension is removed.
  - Quota limitation is approx 5 MB, but can be increased by requesting the `"unlimitedStorage"` permission.
  - Consider using it to store larger amounts of data.

- [**storage.sync**][prop-sync]
  - If syncing is enabled, the data is synced to any Chrome browser that the user is logged into. If disabled, it behaves like `storage.local`.
  - When the browser is offline, Chrome stores the data locally and resumes syncing when it's back online.
  - Quota limitation is 100 KB approx, 8 KB per item
  - Consider using it to preserve user settings across synced browsers. 

{% Aside 'warning' %}

Local and sync storage areas should not store confidential user data because they are not encrypted. We recommend using the `session` storage area, which is not saved on disk for sensitive data.

{% endAside %}

- [**storage.session**][prop-session]
  - Holds small amounts of data in memory for the duration of a browser session.

>   - Not available in content scripts.

Reviewer: This sounded odd to me, so I just tested it and my content script was able to access data in both the sync and local storage areas.

```suggestion
```

Author: @dotproto Did you test the [storage.session](https://developer.chrome.com/docs/extensions/reference/storage/#property-session) in content scripts?

Author: > - Not available in content scripts.

@dotproto Makes sense. How do we want to proceed?

A. Change list item to: Not exposed to content scripts by default, but behavior can be changed by setting [chrome.storage.session.setAccessLevel()](https://developer.chrome.com/docs/extensions/reference/storage/#method-StorageArea-setAccessLevel).
B. Add the example you included.
C. Remove "Not available in content scripts." list item.

Reviewer: Let's run with option A for the moment.

api: storage
---

## Overview {: #overview }

The Storage API provides an extension-specific way to persist user data and state. It's similar to the web platform's storage APIs ([IndexedDB][mdn-indexeddb], [localStorage][mdn-localstorage]), but was designed to meet the storage needs of extensions. The following are a few key features: 

- Content scripts can directly access chrome.storage.
- Stores JSON serializable values as object properties.
- Storage is asynchronous with bulk read and write operations.
- Storage data persists when the user clears the cache and browsing history.
- Available in the service worker, content scripts, and other extension contexts.
- A user's extension settings persist even when using [split incognito][incognito].

> - Offers an exclusive read-only [managed storage area][manifest-storage] for enterprise policies.

Author: PTAL 🧐

{% Details %}
{% DetailsSummary %}
💡 Can extensions use Window.localStorage?

> {% endDetailsSummary %}

Author: It's information that some developers may know but beginner developers don't. I like to pack this additional information into expandable sections so that those who want more information can click for it. It also prevents the document from becoming too bulky.

> {% endDetailsSummary %}

Reviewer 2: Sounds good.

> Even though extensions can access [Window.localStorage][mdn-localstorage] in some contexts (popup and other HTML pages), it is not recommended for the following reasons:

Reviewer 2: Please make sure you're formatting all codeAPI references except those in headings.

> - The extension's service worker cannot access the localStorage API.

Reviewer 2: ```suggestion
- Extension service workers cannot access the web platform's [`localStorage` interface](https://developer.mozilla.org/docs/Web/API/Window/localStorage).
```

Author: Not sure I agree with this change since the extension can only have a single service worker. It implies that an extension can have more than one. Hmm... 🤔

Reviewer 3: Glad I found this. I'm currently attempting to update my chrome extension and have found that instructions related to using Storage API could use some updating. I implemented the options example found here https://developer.chrome.com/docs/extensions/reference/storage/#synchronous-response-to-storage-updates but it generated an error

I go to the options page and it gives me a page from 2018 https://developer.chrome.com/docs/extensions/reference/storage/#synchronous-response-to-storage-updates

Also I agree with @AmySteam that sentence does imply that there can be more than one service worker.

Finally have you all consider full examples showing implementations of the various API?

Author: @AoverK Thanks for your comments. 

Can you try out the code provided in this PR? Here's a link [synchronous-response-to-storage-updates](https://deploy-preview-3961--developer-chrome-com.netlify.app/docs/extensions/reference/storage#synchronous-response-to-storage-updates). If continue running into problems, can you file an issue with more details?

Great suggestion! Actually, we do have plans to create samples for each API :)

> - Content scripts share the localStorage of the host page.

Reviewer 2: ```suggestion
- Content scripts share the `localStorage` of the host page.
```

> - Data saved to localStorage is lost when the user clears their browsing history.

Reviewer 2: ```suggestion
- Data saved to `localStorage` is lost when the user clears their browsing history.
```

{% endDetails %}

### Storage areas

The Storage API is divided into the following four buckets ("storage areas"): 

> - [**storage.local**][prop-local]

Reviewer 2: ```suggestion
[`storage.local`][prop-local]
```

Reviewer 2: Please repeat throughout the list.
* Try to avoid nested bullets or numbered lists.
* That this is a separate line and in a different font sets it off sufficiently enough to make the bullet superfluous.

I'm not sure I would have done bulleted lists here at all. The formatting for that is below (in case you don't know it). GitHub doesn't render this, but our site does.

[`storage.local`][prop-local]
: Stores data locally. The data is cleared when the extension is removed. The quota limitation...

Author: Gotcha 👍

  - Data is stored locally, which is cleared when the extension is removed.
  - Quota limitation is approx 5 MB, but can be increased by requesting the `"unlimitedStorage"` permission.
  - Consider using it to store larger amounts of data.

- [**storage.sync**][prop-sync]
  - If syncing is enabled, the data is synced to any Chrome browser that the user is logged into. If disabled, it behaves like `storage.local`.
  - When the browser is offline, Chrome stores the data locally and resumes syncing when it's back online.
  - Quota limitation is 100 KB approx, 8 KB per item.
  - Consider using it to preserve user settings across synced browsers. 

{% Aside 'warning' %}

> Local and sync storage areas should not store confidential user data because they are not encrypted at rest. When working with sensitive data, consider using the `session` storage area to hold values in memory until the browser is shut down.

Reviewer 2: What does 'not encrypted at rest' mean?

Author: It's a typo. I'm removing it.

{% endAside %}

- [**storage.session**][prop-session]
  - Holds data in memory for the duration of a browser session.

>   - By default, not exposed to content scripts, but behavior can be changed by setting [chrome.storage.session.setAccessLevel()][method-access-level].

Reviewer 2: ```suggestion
  - By default, not exposed to content scripts, but behavior can be changed by setting [`chrome.storage.session.setAccessLevel()`][method-access-level].
```

  - Quota limitation is 1 MB approx.
  - Consider using it to store global variables across service worker runs.

- [**storage.managed**][prop-managed]
  - Administrator can use a [schema][manifest-storage] and enterprise policies to configure a supporting extension's settings in a managed environment.
  - Storage is read-only.

## Manifest {: #manifest}

> To use the storage API, you must declare the `"storage"` permission in the extension

Reviewer 2: ```suggestion
To use the storage API, declare the `"storage"` permission in the extension
```

Reviewer 2: 'You must' is implied well enough by the context. This is an application of "omit needless words." (rule II-17)

Author: 👍

}
```

## Usage {: #usage }

The following samples demonstrate the `local`, `sync`, and

> `sessions` storage areas:

Reviewer 4: Nit: Should this say `session`?

Author: 😅 Whoops! Missed that, thanks! 👍

> <web-tabs>

Reviewer 2: What is this? Examples should avoid using things that require explanation, but are unrelated to the thing being demonstrated.

Author: I'm confused, can you elaborate a bit more? 🤔

Reviewer 2: What is `<web-tabs>`?

Author: It's a component, https://developer.chrome.com/docs/handbook/components/#tabs

<img width="750" alt="Screen Shot 2022-11-28 at 11 13 33" src="https://user-images.githubusercontent.com/37001393/204339919-4f7bdf7c-3d92-4b0f-a050-0fc00e95b42a.png">

[Usage - preview](https://deploy-preview-3961--developer-chrome-com.netlify.app/docs/extensions/reference/storage#usage)

  <web-tab title="storage.local">

  ```js
  chrome.storage.local.set({ key: value }).then(() => {
    console.log("Value is set to " + value);
  });

  chrome.storage.local.get(["key"]).then((result) => {
    console.log("Value currently is " + result.key);
  });
  ```

  </web-tab>
  <web-tab title="storage.sync">

  ```js
  chrome.storage.sync.set({ key: value }).then(() => {
    console.log("Value is set to " + value);
  });

  chrome.storage.sync.get(["key"]).then((result) => {
    console.log("Value currently is " + result.key);
  });
  ```

  </web-tab>
  <web-tab title="storage.session">

  ```js
  chrome.storage.session.set({ key: value }).then(() => {
    console.log("Value is set to " + value);
  });

  chrome.storage.session.get(["key"]).then((result) => {
    console.log("Value currently is " + result.key);
  });
  ```

  </web-tab>
</web-tabs>

To learn more about the `managed` storage area, see [Manifest for storage areas][manifest-storage].

## Storage and throttling limits {: #storage-and-throttling-limits}

> The Storage API is not a big truck. It's a series of tubes. And if you don't understand, those
> tubes can be filled, and if they are filled when you put your message in, it gets in line, and it's
> going to be delayed by anyone that puts into that tube enormous amounts of material.

Reviewer 2: ```suggestion
Don't think of adding to the Storage API as putting things in a big truck. Think of adding to
storage as being like putting something in a pipe. The pipe may have material in it already, and it
may even be filled. Always assume a delay between when you add to storage and when it is actually
recorded.
```

> For details on storage area limitations and what happens when they are exceed, see the quota information for [`sync`][prop-sync], [`local`][prop-local], and [`session`][prop-session].

Reviewer 4: Should this say exceeded?

Author: Absolutely! Thanks again 👍

## Use cases {: #examples}

> The following sections demonstrate some common use cases for the Storage API.

Reviewer 2: ```suggestion
The following sections demonstrate common use cases for the Storage API.
```

### Synchronous response to storage updates

> To track changes made to storage, you can add a listener to its `onChanged` event. When anything changes in storage, that event will fire. The sample code listens for these changes:

Reviewer 2: ```suggestion
To track changes made to storage, you can add a listener to its `onChanged` event. When anything changes in storage, that event fires. The sample code listens for these changes:
```

Reviewer 2: This is non-obvious, but it will probably blow your mind. Most instances of 'will' in computer documentation can be removed. I worked for 14 years at a place that specified this in its style guide. No one ever noted its absence or filed a bug that they didn't understand.

});
```

We can take this idea even further. In this example, we have an [options page][options-page] that
allows the user to toggle a "debug mode" (implementation not shown here).  The options page immediately saves the new settings to `storage.sync`, and the service worker uses `storage.onChanged` to apply the setting as soon as possible.

{% Label %}options.html:{% endLabel %}

```html
<!-- type="module" allows you to use top level await -->
<script defer src="options.js" type="module"></script>

> <form id="optionsForm">

Reviewer: Non-blocking FYI: HTML attribute values are not case sensitive. As such, you will typically HTML attribute values use snake-case while JavaScript variables are typically camelCase.

```js
// In-page cache of the user's options
const options = {};
const optionsForm = document.getElementById("optionsForm");

// Initialize the form with the user's option settings
const data = await chrome.storage.sync.get("options");
Object.assign(options, data.options);

> optionsForm.debug.checked = Boolean(options.debug);

Reviewer: I'm inclined to move these 3 lines below the event listener registration since registering a listener is synchronous, but fetching data from storage is async. We should prepare as much of the application as reasonably possible on the first turn of the event loop. Also, calling `optionsForm.debug.checked = Boolean(options.debug)` will not trigger a change event.

Author: Will do 👍

// Immediately persist options changes
optionsForm.debug.addEventListener("change", (event) => {
  options.debug = event.target.checked;
  chrome.storage.sync.set({ options });
});

> ```

Reviewer: We should declare a `setDebugMode` function in the background.js code sample. Folks sometimes get confused when they see functions like this that aren't declared in the code sample and assume that it's a Chrome global or something. Best to eliminate potential confusion.

```js
function setDebugMode() { /* ... */ }
// or
const setDebugMode = () => { /* ... */ };

Author: Sounds good 👍

});
```

### Asynchronous preload from storage {: #asynchronous-preload-from-storage}

Since service workers are not always running, Manifest V3 extensions sometimes need to
asynchronously load data from storage before they execute their event handlers. To do this, the
following snippet uses an async `action.onClicked` event handler that waits for the `storageCache`
global to be populated before executing its logic.

{% Label %}background.js:{% endLabel %}

> ```js

Reviewer: Thanks for cleaning up this code sample. It was well overdue.

  } catch (e) {
    // Handle error that occurred during storage initialization.
  }

  // Normal action handler logic.
  storageCache.count++;
  storageCache.lastTabId = tab.id;
  chrome.storage.sync.set(storageCache);
});
```

## Extension examples

To see other demos of the Storage API, explore any of the following examples:

- [Global search extension][gh-global-context-search].
- [Water alarm extension][gh-water-alarm].

> 

Reviewer: Remove extra newline.

```suggestion
```
